pyroute2.ndb/pr2modules/ndb/objects/interface.py

```diff
@@ -613,6 +613,30 @@ def complete_key(self, key):
 
         return super(Interface, self).complete_key(ret_key)
 
+    def is_peer(self, other):
+        '''Evaluate whether the given interface "points at" this one.'''
+        other_link = other.get('link')
+
+        if other_link != self['index']:
+            return False
+
+        other_link_netnsid = other.get('link_netnsid')
+        if other_link_netnsid is not None:
+            self_source = self.sources[self['target']]
```

Comment:
This code block doesn't work as expected.

Assume we have three namespaces: main, ns1, ns2
And two veth pairs: main:veth1 + ns1:veth1p, main:veth2 + ns2:veth2p

Now on the kernel 5.11 I get following values for `link_netnsid`:

```
main (!netns)
  veth1['link_netnsid'] == 0
  veth2['link_netnsid'] == 0
ns1 (netns='/var/run/netns/ns1')
  veth1p['link_netnsid'] == 0
ns2 (netns='/var/run/netns/ns2')
  veth2p['link_netnsid'] == 1
```

Why? No idea. Probably it works better on later kernel, but a method used in the core functionality like rollbacks should work on older kernels as well.

I really like the idea of the method, and will continue to debug. @centromere if you have any ideas on that — pls provide comments or PR:s

I replace the method with a stub code that simply compares peer['link'] to host['index'], but later we should get it working properly.

Reply:
Namespace IDs are themselves namespaced. There is a way to discover the list of NSIDs in another namespace. See [`get_netnsid_from_netnsid`](https://github.com/shemminger/iproute2/blob/9f366536edb5158343152604e82b968be46dbf26/ip/ipnetns.c#L375).

```
# ip netns list-id target-nsid 135
nsid 0 current-nsid unassigned (iproute2 netns name: cni-8f68967f-e070-1ee1-9e12-5e36162d84bb)
```

Reply:
Kernel version 5.4:

```
# ip netns add ns1
# ip netns add ns2
# ip l add veth1 type veth peer veth1p netns ns1
# ip l add veth2 type veth peer veth2p netns ns2
```

```
>>> veth1 = n.interfaces['veth1']
>>> veth2 = n.interfaces['veth2']
>>> veth1p = n.interfaces[{ 'target': 'ns1', 'ifname': 'veth1p' }]
>>> veth2p = n.interfaces[{ 'target': 'ns2', 'ifname': 'veth2p' }]
>>> veth1['link_netnsid']
140
>>> veth2['link_netnsid']
141
>>> veth1p['link_netnsid']
0
>>> veth2p['link_netnsid']
0
>>> veth2p.is_peer(veth2)
True
>>> veth2.is_peer(veth2p)
False
```

This logic is at fault:
https://github.com/centromere/pyroute2/blob/1d07b07bc365c82255b99fe0e44d36ce762a5a9f/pyroute2.ndb/pr2modules/ndb/objects/interface.py#L625-L629

Reply:
Here is an [additional resource](https://unix.stackexchange.com/questions/441876/how-to-find-the-network-namespace-of-a-veth-peer-ifindex).